cdap-gateway/src/main/java/co/cask/cdap/gateway/router/handlers/AuthenticationHandler.java

```diff
@@ -0,0 +1,329 @@
+     *
+     * where TOKEN is the base64 encoded serialized AccessToken instance.
+     */
+    String accessToken = null;
+    if (auth != null) {
+      int idx = auth.trim().indexOf(' ');
+      if (idx < 0) {
+        return TokenState.MISSING;
+      }
+
+      accessToken = auth.substring(idx + 1).trim();
+    }
+    TokenState state = tokenValidator.validate(accessToken);
+
+    if (state.isValid()) {
+      try {
+        AccessTokenTransformer.AccessTokenIdentifierPair tokenPair = tokenTransformer.transform(accessToken);
+        // Update message header
+        request.headers().set(HttpHeaderNames.AUTHORIZATION,
+                              "CDAP-verified " + tokenPair.getAccessTokenIdentifierStr());
+        request.headers().set(Constants.Security.Headers.USER_ID,
+                              tokenPair.getAccessTokenIdentifierObj().getUsername());
+        String clientIP = Networks.getIP(channel.remoteAddress());
+        if (clientIP != null) {
+          request.headers().set(Constants.Security.Headers.USER_IP, clientIP);
+        }
+      } catch (Exception e) {
+        // This shouldn't happen in normal case, since the token is already validated
+        LOG.debug("Exception raised when getting token information from a validate token", e);
+        return TokenState.INVALID;
+      }
+    }
+
+    return state;
+  }
+
+  /**
+   * Gets a {@link JsonArray} of url strings to the authentication server instances.
+   */
+  private JsonArray getAuthenticationURLs() {
+    // If the auth server urls are known via configuration, just use it
+    final JsonArray result = new JsonArray();
+
+    if (!authServerURLs.isEmpty()) {
+      for (String url : authServerURLs) {
+        result.add(new JsonPrimitive(url));
+      }
+      return result;
+    }
+
+    // Use service discovery to get URLs of the auth servers
+    final String protocol = getProtocol(cConf);
+    final int port = getPort(cConf);
+
+    ServiceDiscovered serviceDiscovered = discoveryServiceClient.discover(Constants.Service.EXTERNAL_AUTHENTICATION);
+    addAuthServerUrls(serviceDiscovered, protocol, port, result);
+    if (result.size() > 0) {
+      return result;
+    }
+
+    // For bootstrapping, the service discovery takes time to fill in the cache from ZK, hence use a callback
+    // and a timed future to get the result
+    final SettableFuture<JsonArray> future = SettableFuture.create();
+    Cancellable cancellable = serviceDiscovered.watchChanges(new ServiceDiscovered.ChangeListener() {
+      @Override
+      public void onChange(ServiceDiscovered serviceDiscovered) {
+        addAuthServerUrls(serviceDiscovered, protocol, port, result);
+        if (result.size() > 0) {
+          future.set(result);
+        }
+      }
+    }, Threads.SAME_THREAD_EXECUTOR);
+    try {
+      return future.get(2, TimeUnit.SECONDS);
+    } catch (TimeoutException e) {
+      LOG.debug("No authentication server detected via service discovery");
```

Comment:
shouldn't this be a WARN? Perhaps with a Sampling Logger.

Reply:
That was from the old code. Not sure if there is any particular reason for the log level.

Reply:
Anyway, changed it to WARN level since it shouldn't happen normally.